File: ios/Sources/Omnesis/Photos/Analyzers/OCRAnalyzer.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(Vision)
import Vision

/// Extracts on-screen text via `VNRecognizeTextRequest` — the baseline
/// tier that runs for EVERY photo (backfill and new alike), per the
/// issue's "OCR every image" constraint. Always available on-device;
/// no model download, no capability gate.
public struct OCRAnalyzer: PhotoAnalyzer {
    public let identifier = "ocr"

    public init() {}

    public func isAvailable() async -> Bool {
        true
    }

    public func analyze(_ input: PhotoAnalysisInput) async -> PhotoAnalysisFragment? {
        guard let image = input.image else { return nil }
        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true
        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
        } catch {
            return nil
        }
        let lines = (request.results ?? []).compactMap { $0.topCandidates(1).first?.string }
        guard !lines.isEmpty else { return nil }
        return PhotoAnalysisFragment(textLines: lines)
    }
}
#endif

File: android/feature-photos/src/main/kotlin/dev/omnesis/android/feature/photos/analyzers/OcrAnalyzer.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.feature.photos.analyzers

import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import dev.omnesis.android.feature.photos.PhotoAnalysisFragment
import dev.omnesis.android.feature.photos.PhotoAnalysisInput
import dev.omnesis.android.feature.photos.PhotoAnalyzer
import kotlinx.coroutines.tasks.await

/**
 * OCR via ML Kit's on-device Text Recognition (Latin script) — the baseline
 * tier that runs for EVERY photo (backfill and new alike). Bundled model
 * (statically linked, no download), always available; no capability gate.
 * Mirrors iOS's `OCRAnalyzer` (`VNRecognizeTextRequest`).
 */
class OcrAnalyzer : PhotoAnalyzer {
    override val identifier = "ocr"

    private val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    override suspend fun isAvailable(): Boolean = true

    override suspend fun analyze(input: PhotoAnalysisInput): PhotoAnalysisFragment? {
        val bitmap = input.bitmap ?: return null
        val text = recognizer.process(InputImage.fromBitmap(bitmap, 0)).await()
        val lines = text.textBlocks.flatMap { block -> block.lines.map { it.text } }.filter { it.isNotBlank() }
        if (lines.isEmpty()) return null
        return PhotoAnalysisFragment(textLines = lines)
    }
}
